worblehat-domain/src/main/java/de/codecentric/psd/worblehat/domain/StandardBookService.java

```diff
@@ -75,10 +75,13 @@ public void returnAllBooksByBorrower(String borrowerEmailAddress) {
 
 		Optional<Book> bookFromRepo = bookRepository.findTopByIsbn(isbn);
 
-        if (!bookFromRepo.isPresent() || book.isSameCopy(bookFromRepo.get())) {
-            return Optional.of(bookRepository.save(book));
-        } else
-            return Optional.empty();
+		Book newBook = bookFromRepo.orElseGet(() -> bookRepository.save(book));
+
+		Optional<Book> exisitingBookDifferentEdition = bookFromRepo
+				.filter(b -> b.isSameCopy(book))
+				.map(b -> bookRepository.save(b));
+
+		return Optional.of(exisitingBookDifferentEdition).orElse(Optional.of(newBook));

```

Comment:
Wenn hier jetzt auf jeden Fall ein Buch zurück gegeben wird, dann muss der Rückgabetyp nicht mehr `Optional<Book>` sein.

Reply:
Also so wie ich das verstehe, wird immer noch ein Optional.empty zurück gegeben, wenn der filter was leeres zurück gibt. 

Summa sumarum aber schwieriger zu lesen, als der ursprüngliche Code, IMHO

Reply:
Ich kann mit beiden Lösungen leben :-)

Reply:
exisitingBookDifferentEdition sollte aber noch umbenannt werden in newCopyOfExistingBook.

Ich find's so ehrlich gesagt umständlicher und schwieriger zu lesen. Hängt vermutlich auch mit der Erfahrung zusammen.